types/xdg_shell/wlr_xdg_surface.c

```diff
@@ -347,19 +345,48 @@ static void xdg_surface_handle_surface_commit(struct wl_listener *listener,
 	}
 }
 
+static void xdg_surface_handle_surface_precommit(struct wl_listener *listener,
+		void *data) {
+	struct wlr_xdg_surface *surface =
+		wl_container_of(listener, surface, surface_precommit);
+	struct wlr_surface_state *pending = data;
+
+	struct wlr_surface_state_addon *addon = calloc(1, sizeof(*addon));
+	struct wlr_xdg_surface_state *state = calloc(1, sizeof(*state));
+	if (!addon || !state) {
+		free(addon);
+		free(state);
+		return;
+	}
+	memcpy(state, &surface->pending, sizeof(*state));
```

Comment:
Why not make it a pointer?

Reply:
Sorry, not sure what do you mean. `wlr_xdg_surface.pending` is used to store, well, the pending state, just like `wlr_surface.pending`. It has no reason to be a pointer.